File: sem3/PP-scala/laby3/src/main/scala/zad5.scala
object zad5 {
  def palindrome[A](xs: List[A]): Boolean = {
    def reverse_list(ys: List[A]): List[A] = {
      ys match {
        case Nil => Nil
        case head :: tail => reverse_list(tail) :+ head
      }
    }
    xs == reverse_list(xs)
  }
  def main(args: Array[String]): Unit = {
    val list1 = List('a', 'l', 'a')
    val list2 = List('a', 'b', 'c')

    println(palindrome(list1))
    println(palindrome(list2))
  }
}

File: sem3/PP-ocaml/lab3/zad5/zad5.ml
(**ZAD 5*)
(**ocamlc -o zad5 zad1_5/cwiki1_5.ml*)
let print_bool b =
  match b with
  | true -> print_string "true"
  | false -> print_string "false"

let rec palindrome lst =
  let rec reverse_list lst =
    match lst with
    | [] -> []
    | hd::tl -> (reverse_list tl) @ [hd]
  in lst = reverse_list lst

let rec palindrome2 lst = 
  let rec reverse_list2 lst = 
    if lst = [] then []
    else (reverse_list2 (List.tl lst)) @ [(List.hd) lst]
  in lst = reverse_list2 lst

let is_palindrome1 = palindrome ['a'; 'l'; 'a'];;
let is_palindrome2 = palindrome ['a'; 'b'; 'c'];;
let is_palindrome3 = palindrome ['a'; 'l'; 'l'; 'a'];;

print_bool is_palindrome1;;
print_newline ();;
print_bool is_palindrome2;;
print_newline ();;
print_bool is_palindrome3;;
print_newline ();;
